File: crypto/src/crc32.rs
//! CRC-32 (IEEE 802.3 polynomial 0xEDB88320 reflected).

/// CRC-32 lookup table (256 entries, reflected polynomial).
#[cfg(feature = "crc32-table")]
const CRC32_TABLE: [u32; 256] = {
    let mut table = [0u32; 256];
    let mut i = 0u32;
    while i < 256 {
        let mut crc = i;
        let mut j = 0;
        while j < 8 {
            if crc & 1 != 0 {
                crc = (crc >> 1) ^ 0xEDB88320;
            } else {
                crc >>= 1;
            }
            j += 1;
        }
        table[i as usize] = crc;
        i += 1;
    }
    table
};

/// Compute CRC-32 over a byte slice (table-based, fast).
#[cfg(feature = "crc32-table")]
pub fn crc32(data: &[u8]) -> u32 {
    let mut crc = 0xFFFFFFFFu32;
    for &byte in data {
        let idx = ((crc ^ byte as u32) & 0xFF) as usize;
        crc = (crc >> 8) ^ CRC32_TABLE[idx];
    }
    crc ^ 0xFFFFFFFF
}

/// Compute CRC-32 over a byte slice (no table, smaller code size).
#[cfg(not(feature = "crc32-table"))]
pub fn crc32(data: &[u8]) -> u32 {
    let mut crc = 0xFFFFFFFFu32;
    for &byte in data {
        crc ^= byte as u32;
        for _ in 0..8 {
            if crc & 1 != 0 {
                crc = (crc >> 1) ^ 0xEDB88320;
            } else {
                crc >>= 1;
            }
        }
    }
    crc ^ 0xFFFFFFFF
}

/// Incrementally update a running CRC-32 value.
pub fn crc32_update(crc: u32, data: &[u8]) -> u32 {
    let mut c = crc ^ 0xFFFFFFFF;
    for &byte in data {
        #[cfg(feature = "crc32-table")]
        {
            let idx = ((c ^ byte as u32) & 0xFF) as usize;
            c = (c >> 8) ^ CRC32_TABLE[idx];
        }
        #[cfg(not(feature = "crc32-table"))]
        {
            c ^= byte as u32;
            for _ in 0..8 {
                if c & 1 != 0 {
                    c = (c >> 1) ^ 0xEDB88320;
                } else {
                    c >>= 1;
                }
            }
        }
    }
    c ^ 0xFFFFFFFF
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_crc32_empty() {
        assert_eq!(crc32(b""), 0x00000000);
    }

    #[test]
    fn test_crc32_check_value() {
        // The CRC-32 of "123456789" is 0xCBF43926 (standard check value)
        assert_eq!(crc32(b"123456789"), 0xCBF43926);
    }

    #[test]
    fn test_crc32_single_byte() {
        assert_eq!(crc32(b"A"), 0xD3D99E8B);
    }

    #[test]
    fn test_crc32_incremental() {
        let full = crc32(b"hello world");
        let partial = crc32_update(crc32(b"hello "), b"world");
        assert_eq!(full, partial);
    }
}

File: crypto/src/crc32.c
/*
 * crc32.c — CRC-32 (IEEE 802.3, reflected polynomial 0xEDB88320).
 *
 * Define GEN_CRC32_NO_TABLE to use bit-by-bit fallback (saves 1 KB).
 * Freestanding.
 */

#include "gen_crypto.h"

#ifndef GEN_CRC32_NO_TABLE

/* ── Table-based (fast) ─────────────────────────────────── */

/*
 * Build the CRC-32 lookup table at program initialization.
 * Alternatively, a pre-computed table could be used as a static const.
 */
static uint32_t crc32_table[256];
static int crc32_table_ready = 0;

static void crc32_build_table(void)
{
    if (crc32_table_ready) return;
    for (uint32_t i = 0; i < 256; i++) {
        uint32_t crc = i;
        for (int j = 0; j < 8; j++) {
            if (crc & 1)
                crc = (crc >> 1) ^ 0xEDB88320U;
            else
                crc >>= 1;
        }
        crc32_table[i] = crc;
    }
    crc32_table_ready = 1;
}

static uint32_t crc32_compute(uint32_t crc, const uint8_t *data, size_t len)
{
    crc32_build_table();
    for (size_t i = 0; i < len; i++) {
        uint32_t idx = (crc ^ data[i]) & 0xFFU;
        crc = (crc >> 8) ^ crc32_table[idx];
    }
    return crc;
}

#else /* GEN_CRC32_NO_TABLE */

/* ── Bit-by-bit (small) ─────────────────────────────────── */

static uint32_t crc32_compute(uint32_t crc, const uint8_t *data, size_t len)
{
    for (size_t i = 0; i < len; i++) {
        crc ^= data[i];
        for (int j = 0; j < 8; j++) {
            if (crc & 1)
                crc = (crc >> 1) ^ 0xEDB88320U;
            else
                crc >>= 1;
        }
    }
    return crc;
}

#endif /* GEN_CRC32_NO_TABLE */

uint32_t gen_crc32(const uint8_t *data, size_t len)
{
    return crc32_compute(0xFFFFFFFFU, data, len) ^ 0xFFFFFFFFU;
}

uint32_t gen_crc32_update(uint32_t crc, const uint8_t *data, size_t len)
{
    return crc32_compute(crc ^ 0xFFFFFFFFU, data, len) ^ 0xFFFFFFFFU;
}
